src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,411 @@
+
+const std = @import("std");
+const math = std.math;
+const mem = std.mem;
+const testing = std.testing;
+const ArrayList = std.ArrayList;
+
+const tersets = @import("../tersets.zig");
+const Error = tersets.Error;
+
+/// Segment represented by the points: (start_time, start_value) and (end_time, end_value).
+/// The `start_time` and `end_time` represent indices of the time series.
+const Segment = struct {
+    start_time: usize,
+    start_value: f64,
+    end_time: usize,
+    end_value: f64,
+};
+
+/// LinearFunction represented by the function y = slope*x+intercept.
+/// It uses f80 to improve numerical stability.
+const LinearFunction = struct {
+    slope: f80,
+    intercept: f80,
+};
+
+/// Converts `value`:usize to f80
+fn usizeToF80(value: usize) f80 {
+    return @as(f80, @floatFromInt(value));
+}
+
+/// Computes the intercept coefficient of a line that passes through the point
+/// (x_time, y_value) and with the given slope coefficient.
+fn computeInterceptCoefficient(slope: f80, x_time: usize, y_value: f64) f80 {
+    return y_value - slope * usizeToF80(x_time);
+}
+
+/// Updates the linear function coeficients in `line` that passes throught the two points
+/// of the `segment`. The line is swinged down or up based on the `error_bound`.
+/// If `error_bound` is negative, `line` is swing down. It is swing up otherwise.
+fn updateSwingLine(segment: Segment, line: *LinearFunction, error_bound: f32) void {
+    if (segment.end_time != segment.start_time) {
+        const duration: f80 = @floatFromInt(segment.end_time - segment.start_time);
+        line.slope = (segment.end_value + error_bound - segment.start_value) / duration;
+        line.intercept = segment.start_value - line.slope * usizeToF80(segment.start_time);
+    } else {
+        line.slope = 0.0;
+        line.intercept = segment.start_value;
+    }
+}
+
+/// Evaluate the linear function `line` at `x_time`.
+fn evaluateLineAtTime(line: LinearFunction, x_time: usize) f64 {
+    return @floatCast(line.slope * usizeToF80(x_time) + line.intercept);
+}
+
+/// Append `value` to `compressed_values`.
+fn appendValue(value: f64, compressed_values: *ArrayList(u8)) !void {
+    const valueAsBytes: [8]u8 = @bitCast(value);
+    try compressed_values.appendSlice(valueAsBytes[0..]);
+}
+
+/// Append `index` to `compressed_values`.
+fn appendIndex(index: usize, compressed_values: *ArrayList(u8)) !void {
+    const indexAsFloat: [8]u8 = @bitCast(@as(f64, @floatFromInt(index)));
+    try compressed_values.appendSlice(indexAsFloat[0..]);
+}
+
+/// Computes the numerator of the slope derivate as in Eq. 6.
+fn computeSlopeDerivate(segment: Segment) f80 {
+    return (segment.end_value - segment.start_value) * usizeToF80(segment.end_time - segment.start_time);
+}
+
+/// Compress `uncompressed_values` within `error_bound` using "Swing Filter"
+/// and write the result to `compressed_values`. If an error occurs it is returned.
+pub fn compress_swing(
```

Comment:
Sine the Zig style guide states that "If x is otherwise callable, then x should be camelCase." I think our `compress_x` should actually be named `compressX`. This is my mistake, although it should be easy to automatically renaming.

Reply:
I did not notice this before. I will add a TODO for a later PR.